Integration testing strategy (#61)
buerokratt/Cross-functional-requirements: docs/technical/integration-testing-strategy.md

Integration tests are not to be confused with "unit tests" which make assertions of specific code modules and therefore operate at the code level, before the system component has been deployed and integrated. By the time a component is exposed to integration tests, it should have already passed all of its unit tests.

## Tests and Assertions

This section will outline the integration tests that will be created and what assertions they will make. This only outlines the initial "happy path" tests that are required to build assurances that the integrated system is operating correctly.

In passing these tests, all system components (Mock Bot, DMR, Mock Classifier and CentOps) will have been exercised. Additional tests which check edge cases and error cases can be added later.

### Classify a message

This test asserts that when DMR receives an unclassified message which is correctly formed, it is able to classify it and reply to the calling participant with the classification. This is steps 1-4 from [Milestones.md](https://github.com/buerokratt/Project-Documentation-Management/blob/main/Milestones.md) and represents milestone 1.

> Input: Test requests that mock bot generates an unclassified message with all appropriate headers and submits it to DMR.

Reviewer: There might be some setup with enrolling the bot(s) needed by tests with CentOps.

Author: Added setup steps.

Input: Test requests that mock bot generates a classified message with all appropriate headers and submits it to DMR.

Expected output: DMR forwards the message to the correct mock bot. Test asserts that the message was received and is in the expected format.

## How to make assertions on HTTP calls

The architecture of the Bürokratt system operates on a series of chained HTTP requests and the final step of any interaction is for a mock bot to receive a HTTP request from DMR.

In order to assert that these HTTP requests are being received, we can use some of the existing endpoints that are included in the mock bot to create chats and post messages to a chat (thereby sending it to DMR).

There may be some minor extensions required to mock bot to ensure that when DMR responds back to the bot, the message is added to the Chat (currently it is disposed). This will enable us to simply get the chat and inspect the contents which should contain both the initial message out to DMR and the classified response back from DMR.

## Platform and Tooling

All integration tests will be written using C# and XUnit. This is a consistent choice with the unit tests for the Bürokratt system.

> The tests will use an `appsettings.json` and `appsettings.developments.json` files to store configuration information for accessing components (endpoints, api keys etc).

Reviewer: Will this always be the case?  I was wondering if some of the settings (api keys) might come from environment variables when running in a pipeline.  I'm probably misunderstanding here.

Author: Fixed, thanks

Developers will be able to run tests locally using the `dotnet test` command.

## Pipelines

The goal of running tests as part of pipelines that the outcome of a test can be used a gate for progressing code to the next stage. For example if a test fails, then the automated progression of code from dev to production can be halted.

> The challenge with running integration tests as part of a pipeline is that the code must be available in its integrated form in order for the test to work, so to a certain extent the pipline must have progressed before integration tests can run.

Reviewer: ```suggestion
The challenge with running integration tests as part of a pipeline is that the code must be available in its integrated form in order for the test to work, so to a certain extent the pipeline must have progressed before integration tests can run.
```

> The challenge with running integration tests as part of a pipeline is that the code must be available in its integrated form in order for the test to work, so to a certain extent the pipline must have progressed before integration tests can run.

Reviewer 2: This was a little difficult to read so here my suggestion on how to reword it
```suggestion
The challenge with running integration tests as part of a pipeline is that the code must be deployed to the `dev` environment (our first integrated environment), meaning that it needs to have merged into the `main` branch already.
```

Author: I'm not sure if the suggestion is quite right either because we can run the tests with Docker Compose. I'll have a look and come up with some better words.

Author: I ended up removing the paragraph entirely as it was not really adding anything

>
> https://github.com/buerokratt/DMR

### Docker via code repository CI pipeline

One advantage of all the code components being hosted as containers is that we can easily run these containers locally via Docker, they do not necessarily need to be deployed to Azure to be "integrated".

There will be a new pipeline added to each code repository which uses [Docker Compose](https://docs.docker.com/compose/) to pull and run the containers that published as part of the `ci-build-publish-main.yml` and execute the integration tests against these locally running containers. 

This pipeline will be a requirement for PR's and merges (a "gate") just like the other pipelines in each code repository.

While this does not give us full assurance that the components integrate in their deployed state, it will enable us to catch integration issues early on.

### Deployed components via infrastructure CD pipeline

In addition to running the integration tests as part of the code repository CI pipeline, the tests will be run again against the deployed version of the code in the Azure Dev environment.

This will use environment data from Terraform to exercise the live components.

> Sucessfull completion of the test against the Dev environment will be a gate for the code progressing to the prod environment.

Reviewer: ```suggestion
Successful completion of the test against the Dev environment will be a gate for the code progressing to the prod environment.
```
